flag_engine/features/models.py

```diff
@@ -19,41 +21,75 @@ def __hash__(self):
         return hash(self.id)
 
 
-@dataclass
-class MultivariateFeatureOptionModel:
+class MultivariateFeatureOptionModel(BaseModel):
     value: typing.Any
     id: int = None
 
 
-@dataclass
-class MultivariateFeatureStateValueModel:
+class MultivariateFeatureStateValueModel(BaseModel):
     multivariate_feature_option: MultivariateFeatureOptionModel
-    percentage_allocation: float
+    percentage_allocation: confloat(ge=0, le=100)
     id: int = None
-    mv_fs_value_uuid: str = field(default_factory=uuid.uuid4)
+    mv_fs_value_uuid: UUID4 = Field(default_factory=uuid.uuid4)
 
 
-@dataclass
-class FeatureSegmentModel:
+class FeatureSegmentModel(BaseModel):
     priority: int = None
 
 
-@dataclass
-class FeatureStateModel:
+class MultivariateFeatureStateValueList(
+    BaseCollectionModel[MultivariateFeatureStateValueModel]
+):
+    @classmethod
+    def __get_validators__(
+        cls,
+    ) -> typing.Generator[typing.Callable[..., typing.Any], None, None]:
+        yield cls.validate
+        yield cls.ensure_correct_percentage_allocations
+
+    @classmethod
+    def ensure_correct_percentage_allocations(
+        cls,
+        value: typing.List[MultivariateFeatureStateValueModel],
+    ) -> typing.List[MultivariateFeatureStateValueModel]:
+        if (
+            sum(
+                multivariate_feature_state.percentage_allocation
+                for multivariate_feature_state in value
+            )
+            > 100
+        ):
+            raise InvalidPercentageAllocation(
+                "Total percentage allocation for feature must be less than 100 percent"
+            )
+        return value
+
+    def append(
+        self,
+        multivariate_feature_state_value: MultivariateFeatureStateValueModel,
+    ) -> None:
+        super().append(multivariate_feature_state_value)
```

Comment:
I think this is a bit odd that validation happens after the object is modified, e.g. if append raises an error on `some object`  I would not expect that object to be modified  - Does that make sense?

Reply:
Agreed, fixed for both custom collections.